### PR TITLE
Implement DNS cache recycling

### PR DESCRIPTION
**Pull request by DL6ER**
# What does this implement/fix?

This PR completes what was started in #1694 - memory recycling. The issue at hand was caused by adding the facility to recycle DNS cache records but then never actually doing it. The missing bits are added in this PR.

This fixes an issue described on Discourse (see link below): After a certain time has passed, clients and domains may be recycled, however the cache records saying that "clientID=123,domainID=234 -> blocked" remained. This was harmless *until* a new (possibly but not necessarily the old) client/domain combination in exactly this configuration appeared which then lead to a cached reply which was actually intended for something entirely different.

This is fixed by recycling cache records as soon as they are not referenced by any query. This is similar to how we recycle clients and domains, too - we remove them when the last query referencing them disappeared due to aging (> 24h).

---

**Related issue or feature (if applicable):** N/A

**Pull request in [docs](https://github.com/pi-hole/docs) with documentation (if applicable):** N/A

---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
2. I have commented my proposed changes within the code.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

## Checklist:

- [x] The code change is tested and works locally.
- [x] I based my code and PRs against the repositories `developmental` branch.
- [x] I [signed off](https://docs.pi-hole.net/guides/github/how-to-signoff/) all commits. Pi-hole enforces the [DCO](https://docs.pi-hole.net/guides/github/dco/) for all contributions
- [x] I [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) all my commits. Pi-hole requires signatures to verify authorship
- [x] I have read the above and my PR is ready for review.

### CODE REVIEW

**Comment by DL6ER** [comment]
Confirmed working by [OP on Discourse](https://discourse.pi-hole.net/t/pi-hole-blocking-domain-it-shouldnt/67111/24?u=dl6er)

File: src/gc.c [hunk 1]
```diff
@@ -378,17 +377,17 @@ void runGC(const time_t now, time_t *lastGCrun, const bool flush)
 
 		// Update reply counters
 		counters->reply[query->reply]--;
-		log_debug(DEBUG_GC, "reply type %d removed (GC), ID = %d, new count = %d", query->reply, query->id, counters->reply[query->reply]);
+		log_debug(DEBUG_STATUS, "reply type %d removed (GC), ID = %d, new count = %d", query->reply, query->id, counters->reply[query->reply]);
```
Comment (yubiuser) [inline src/gc.c:380]: Why did you change all the instances from ` DEBUG_GC > DEBUG_STATUS`?

Here it even talks about the garbage collection.

Comment (DL6ER) [reply src/gc.c:380]: Because the status debug flag is meant specifically to log every single counter change (for every query!) so it's very very verbose. The GC debug flag prints much fewer lines with more condensed content. As these lines make no sense without DEBUG_STATUS being set at the same time (for context), I moved them over there

File: src/log.c [hunk 2]
```diff
@@ -482,6 +482,7 @@ void log_counter_info(void)
 	log_info(" -> Unknown DNS queries: %i", counters->status[QUERY_UNKNOWN]);
 	log_info(" -> Unique domains: %i", counters->domains);
 	log_info(" -> Unique clients: %i", counters->clients);
+	log_info(" -> DNS cache records: %i", counters->dns_cache_size);
```
Comment (yubiuser) [inline src/log.c:485]: What does "DNS cache records" tell us? How does it relate to the imported long-term data?

```
2024-01-04 22:11:36.590 [430399M] INFO: Imported 8396 queries from the long-term database
2024-01-04 22:11:36.598 [430399M] INFO:  -> Total DNS queries: 8396
2024-01-04 22:11:36.599 [430399M] INFO:  -> Cached DNS queries: 5696
2024-01-04 22:11:36.599 [430399M] INFO:  -> Forwarded DNS queries: 1889
2024-01-04 22:11:36.599 [430399M] INFO:  -> Blocked DNS queries: 35
2024-01-04 22:11:36.599 [430399M] INFO:  -> Unknown DNS queries: 4
2024-01-04 22:11:36.599 [430399M] INFO:  -> Unique domains: 583
2024-01-04 22:11:36.599 [430399M] INFO:  -> Unique clients: 3
2024-01-04 22:11:36.599 [430399M] INFO:  -> DNS cache records: 937
2024-01-04 22:11:36.599 [430399M] INFO:  -> Known forward destinations: 2

```

Comment (DL6ER) [reply src/log.c:485]: Every unique client/domain combination creates a unique DNS cache record. Two queries for the same domain use one cache record. Two queries for the same domain by three clients create three cache records (because some clients may be blocking a domain while others don't due to groups). Thia is just yet another metric used by FTL for quite some time and which seemed useful to include here.

**Review by yubiuser — APPROVED** [review]
